### ISSUE TITLE
[legacy versions] how do we support old versions

### ISSUE THREAD

**Issue by chenqin** [dmlc/xgboost#4734]
As xgboost is getting into production with critical use cases, I would like to open a discussion on how we support old versions. This is important because

1. changes from different versions may change model performance 
2. change from different versions may require code change or dependency change

Both of which will put extra work and more importantly a risk factor to platform teams (ML as a service) 

Here is similar discussion from another Apache project.
http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/DISCUSS-Time-based-releases-in-Flink-td15386.html#a15394

@hcho3 @CodingCat

**Comment by CodingCat** [comment]
we have achieved the agreement on how to support old version and how to achieve backward compatibility. check https://xgboost.readthedocs.io/en/latest/contrib/release.html

In most of cases (not releasing a major version), the following two cases is not to be there unless in a form of bug

> 1. changes from different versions may change model performance
> 2. change from different versions may require code change or dependency change

**Comment by hcho3** [comment]
@CodingCat 

> the following two cases are not to be there unless in a form of bug

Are we sure that we won't have situations like 0.82 -> 0.90 upgrade?

@chenqin Can you be more specific? Do you mean that we should support old releases, e.g. backport bug fixes to previous releases to some time? See #4389, where we previously decided that we didn't have enough manpower and resources to maintain legacy releases.

**Comment by CodingCat** [comment]
> Are we sure that we won't have situations like 0.82 -> 0.90 upgrade?

>> 1. changes from different versions may change model performance 

isn't this a bug? I don't think any new change would shift the output of algorithm significantly given the current status of algorithms (cpu at least, gpu maybe) 

>> 2. change from different versions may require code change or dependency change

this breaks into two parts

(1) code change: no we will try our best to avoid the case like https://github.com/dmlc/xgboost/pull/4349 in future (the major reason for me to put a simplified solution like that is we were in 0.x where a lot of flexibilities exist), if anything similar appeared in future, we should consider behavior consistency

(2) dependency change: I cannot find it, but I think in some thread, we decided to drop the binary compatibility guarantee in the end. Not sure if it is worth bringing back the discussion here. On the other side, as a project maintainer, we should be strategic for major dependency upgrading (like Spark). For instance, what's the right time to move forward from 2.4 to 3.0. (if Spark 3 was going to release right before next Spark Summit, shall we wait...if Spark 3 is releasing next month, shall we upgrade exclusively or compatibly, given the risk that the first major release of spark is always `full of surprise`)

**Comment by chenqin** [comment]
1. changes from different versions may change model performance

Only be vocal about reproducibility. I only know one case from my side, could be a bug. I asked people familiar with internal better, what I heard was there were placed worth looking into. Take 1) back.

2. change from different versions may require code change or dependency change
>dependency change

I tried to build against master and launch experiment job. Found job failed because we switch from scala 2.11 to 2.12 after 0.9 and spark were built against 2.11. I guess that shouldn't be a problem for myself. But it might cause some pain for certain group.

>code change

The idea behind is if people hire contractor/employee already left(non tech company) built on top of this project. Those work should compile and run without worry about bugs not fixed for a reasonable period of time (upgrade friendly wise) Declaimer: I am not sure if this apply to our audience use case at large.

https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/annotation/Public.html
